Make UBL a complete singleton

thinkyhead opened the pull request:
- Move most UBL-related static functions into the `unified_bed_leveling` class.
- Add 'user_canceled' for repeated use by `G26`.
- Move G26 and G29 arguments into the class.
- Simplify `mesh_index_to_xpos` and `mesh_index_to_ypos` interface.

Also:
- Cleanups for `fastio.h`

thinkyhead committed:
Clean up whitespace

thinkyhead committed:
Single user_canceled function in G26

Roxy-3D commented:
That is good...   But there is a similar construct in G29 several places.  The G26 stuff is easier to swap out because it doesn't care about normal clicks.   Just Press & Hold's.   

But in the G29 P4 (and P2) cases, it is possible the user is doing a 'Click'.  Or it is possible the user is doing a 'Press and Hold' to cancel.   It would be good to compress that logic too and maybe make this new routine tell us which of the three possibilities happened:
- Nothing
- a single click
- a Press & Hold

thinkyhead commented:
Let's work on a standardized method within the LCD code for this kind of usage.

Or, you can follow the pattern used by those functions which work right now without requiring any debounce logic.

thinkyhead commented:
Let's work on a standardized method within the LCD code for this kind of usage.

Or, we can follow the pattern used by those functions which work _right now_ without requiring any additional debounce logic.

Press-and-hold is something we can certainly add standard code to look for, when desired. And, we can also add an option for the LCD to ignore clicks and let "someone else" deal with them as a standardized behavior. Though, it would be best for the LCD to handle all the details of the button being pressed, debounce, etc., and simply set flags that other code may examine.

Roxy-3D commented:
So this gets moved into ultraLCD.cpp ?   If we do that...  It might make sense to have a few flags we could pass in to control behavior.  It might make sense to tell the routine what is possible just because it could be more responsive on the clicks if it doesn't have to wait around for the release to know if it was a Press and Hold.   

Also...  If this was integrated with and moved into ultraLCD.cpp all of that UBL_has_control_of_lcd can go away too...

thinkyhead commented:
Not necessarily "this moves" but we add proper support into the LCD code for this kind of usage. You would set a flag, such as `lcd_detect_click_and_hold` that would watch for that and respond in some way. The simplest and most universal solution I can imagine simply sets a function which should be called.
```cpp
ultralcd.click_and_hold_callback = ubl_cancel_my_action();
```

Roxy-3D commented:
OK...    So maybe there are two ways to get results...   Maybe the LCD code sets a flag (that probably automatically gets cleared after it is checked).   And maybe it can be told to call_back() into some code if the condition happens.    Right now, neither G26 or G29 is setup for call_back()'s.    But future code probably would like that as an option.    Right now...  Both G26 and G29 could be simplified if they checked a flag that ultralcd.cpp was maintaining...

thinkyhead commented:
Not necessarily "this moves" but we add proper support into the LCD code for this kind of usage. You would set a flag, such as `lcd_detect_click_and_hold` that would watch for that and respond in some way, or…

The simplest and most universal solution I can imagine simply sets a function which should be called.
```cpp
void ubl_cancel_my_action() { action_cancel = true; }
```
…
```cpp
void gcode_G26() {
  ...
  ultralcd.click_and_hold_callback = ubl_cancel_my_action;
  ...
  if (action_cancel) {
    // Cancel this command and exit
  }
  Exit:
  ultralcd.click_and_hold_callback = lcd_goto_babystepping;
}
```

thinkyhead commented:
> they checked a flag that ultralcd.cpp was maintaining...

Let the ultralcd code maintain its internal state, ignorant of the rest of the code. We'll give it callbacks for things like this, so that other code can maintain its own private flags.

Roxy-3D commented:
OK...  I understand.    That will work...

thinkyhead commented:
I mean, ultimately, I would love it if we could work towards a more "component-based architecture" for the high level code. This pattern provides a set of standard events (e.g., "button clicked") and objects simply register to get a callback when that occurs. Every registered object will have its handler called from a list of "listeners." In this way, every object decides how to respond to events, and it doesn't have to know anything about how those events are being generated. It can also be applied to periodic events, such as "at this time, call me back," or "every N interval….", or "after a line segment is queued."

https://en.wikipedia.org/wiki/Observer_pattern

Roxy-3D commented:
Yeah.   That makes sense...     Unfortunately, the LCD code was written thinking the only thing that would ever need the encoder wheel or the LCD Panel was the lcd code...    But that strategy might let us cut through the knot and get everything untangled.

thinkyhead commented:
> the LCD code was written thinking the only thing that would ever need the encoder wheel or the LCD Panel was the lcd code

It isn't especially bad, and it has been moving in a better direction. Remember when every menu handler asked the hardware for the click state and did its own debounce logic? Now we check in just one place and set a flag.

In the same vein, we can have the LCD be a "listener" for its own events. Most of the time, this means entering the menus, etc. When we want it to ignore events and let someone else handle them, we just have the LCD stop being a listener for a while and let someone else receive event callbacks.

This is the kind of stuff we will be able to do better and better as we keep modernizing the code. A centralized broadcaster-listener and smart application of callbacks allows the rest of the code to focus just on its own specific domains.

thinkyhead committed:
Cleanup some fastio code

thinkyhead committed:
Use new attribute macros in more places

thinkyhead committed:
Apply const float & more

thinkyhead committed:
Embed G26/G29 in ubl class, with enhancements

thinkyhead commented:
@Roxy-3D I have some additional proposed UBL changes in this PR — mainly just cleanup / simplification type stuff. Specifically:

- Made inline accessors for the `mesh_index_to_xpos`/`mesh_index_to_ypos` arrays. Adds no overhead, just makes use of these arrays easier to read and type.
- Moved the `G26` and `G29` handlers into the ubl class so they don't have to go through the `ubl` object to get properties or make method calls.
- Moved UBL motion functions into the class also.

Roxy-3D commented:
When Travis says "Ok!" go ahead and merge...   I'll load it up and check out the stuff that changed...

thinkyhead commented:
Ok, I've been over it a lot. I tried not to alter any logic, just cosmetics and namespaces. I hope it holds together!

Roxy-3D commented:
OK...  I'll try to check the code paths where stuff has changed....

bgort commented:
I thought we were headed in the direction of `G26` being a standalone command that could be used to validate any bed leveling system?

Roxy-3D commented:
Oh!  I missed that...

thinkyhead commented:
I don't think that's going to happen very soon, and it still seems to be very much in flux. Get it solid and stable with UBL first, and then we can look at how we want to merge it into the global command pile. I would also like to make sure we have a more standardized system of interfacing with the LCD controller, so there isn't so much knob/button logic in classes devoted to higher-level stuff.

Roxy-3D commented:
OK.  That is fine.   But G26 is stable.   And works as expected.   It is very valuable in mesh leveling schemes because it lets you see how well each mesh point is adjusted.

thinkyhead commented:
Excellent. Everything is going exactly to plan.

landodragon141 commented:
Ohhh my!

bgort commented on Marlin/ubl_motion.cpp:
```diff
@@ -606,19 +606,19 @@
-                    y0 = pgm_read_float(&(ubl.mesh_index_to_ypos[cell_yi  ])),  // 64 byte table lookup avoids mul+add
-                    x1 = pgm_read_float(&(ubl.mesh_index_to_xpos[cell_xi+1])),  // 64 byte table lookup avoids mul+add
-                    y1 = pgm_read_float(&(ubl.mesh_index_to_ypos[cell_yi+1]));  // 64 byte table lookup avoids mul+add
+        const float x0 = pgm_read_float(&(mesh_index_to_xpos[cell_xi  ])),  // 64 byte table lookup avoids mul+add
```
These lines (4) don't work, as mesh_index_to_Xpos is now a function.

thinkyhead replied:
Darn, thought I got 'em all — thanks! #6836

bgort replied:
No worries.  Luckily I saw it when reviewing rather than finding it the hard way....
